base/iterators.jl

```diff
@@ -30,6 +30,44 @@ and_iteratorsize(a, b) = SizeUnknown()
 and_iteratoreltype(iel::T, ::T) where {T} = iel
 and_iteratoreltype(a, b) = EltypeUnknown()
 
+## Reversed-order iteration for arrays and other collections.  Collections
+## should implement start/next/done etcetera if possible/practical.
+"""
+    Reversed(itr)
+
+Given an iterator `itr`, then `Reversed(itr)` is an iterator over the
+same collection but in the reverse order.
+
+Not all iterator types `T` support reverse-order iteration.  If `T`
+doesn't, then iterating over `Reversed(itr::T)` will throw a [`MethodError`](@ref)
+because of the missing [`start`](@ref), [`next`](@ref), and [`done`](@ref)
+methods for `Reversed{T}`.
+"""
+struct Reversed{T}
+    itr::T
+end
+eltype(r::Reversed) = eltype(r.itr)
+length(r::Reversed) = length(r.itr)
+size(r::Reversed) = size(r.itr)
+iteratorsize(r::Reversed) = iteratorsize(r.itr)
+iteratoreltype(r::Reversed) = iteratoreltype(r.itr)
+last(r::Reversed) = first(r.itr) # the first shall be last
+first(r::Reversed) = last(r.itr) # and the last shall be first
+
+# reverse-order array iterators: assumes more-specialized Reversed for eachindex
+@inline start(A::Reversed{<:AbstractArray}) = (itr = Reversed(eachindex(A.itr)); (itr, start(itr)))
+@propagate_inbounds next(A::Reversed{<:AbstractArray}, i) = ((idx, s) = next(i[1], i[2]); (A.itr[idx], (i[1], s)))
+@propagate_inbounds done(A::Reversed{<:AbstractArray}, i) = done(i[1], i[2])
+
+Reversed(R::AbstractRange) = reverse(R) # copying ranges is cheap
+Reversed(G::Generator) = Generator(G.f, Reversed(G.itr))
+Reversed(r::Reversed) = r.itr
+reverse(r::Reversed) = collect(r.itr)
```

Comment:
Same question regarding `collect`. Seems odd for strings. (GitHub ate my previous comment.)

Reply:
Why is this odd for strings?  `Reversed{<:String}` is an iterator over characters, so should `reverse(r)` on it be the same as `reverse(collect(r))`?

Reply:
I would have expected that since the underlying object being iterated is a string, unwrapping the `Reversed` wrapper would give you back the string, just as `reverse(::String)` gives you a string.